### ISSUE TITLE
vmware_guest_facts fails with missing parameter when searching by UUID

### ISSUE THREAD

**Issue by jmcginn13** [ansible/ansible#25713]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
vmware_guest_facts

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel cf30b162a9) last updated 2017/06/13 22:02:47 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/jmcginn/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jmcginn/workspaces/ansible/ansible/lib/ansible
  executable location = /home/jmcginn/workspaces/ansible/ansible/bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```
NOTE: I also reproduced this on the latest 2.3.0 release too

##### CONFIGURATION
Using default configuration with no ansible.cfg

##### OS / ENVIRONMENT
Centos 7
uname -r: 3.10.0-514.16.1.el7.x86_64
ESXi 6

pip show pyvmomi
```
Name: pyvmomi
Version: 6.5.0.2017.5.post1
Summary: VMware vSphere Python SDK
Home-page: https://github.com/vmware/pyvmomi
Author: VMware, Inc.
Author-email: jhu@vmware.com
License: License :: OSI Approved :: Apache Software License
Location: /usr/lib/python2.7/site-packages
Requires: requests, six
```

##### SUMMARY
vmware_guest_facts fails to search by uuid with the following task:
```
 name: Lookup my VM facts
      vmware_guest_facts:
       name: "{{ vm_name }}"
        uuid: "{{ vm_uid }}"
        datacenter: "{{ my_datacenter }}"
        hostname: "{{ vmware_host }}"
        username: "{{ vmware_user }}"
        password: "{{ vmware_password }}"
        validate_certs: no
      register: my_vm
```
**NOTE: I had to include name because of issue [24398](https://github.com/ansible/ansible/issues/24398) **
##### STEPS TO REPRODUCE
Run the above task with a valid instance UUID from a VM in a vsphere cluster

##### EXPECTED RESULTS
See detail facts about the VM with the matching UID

##### ACTUAL RESULTS
Get the following stack track
Get the following stacktrack:
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_xTG64q/ansible_module_vmware_guest_facts.py", line 236, in <module>
    main()
  File "/tmp/ansible_xTG64q/ansible_module_vmware_guest_facts.py", line 223, in main
    uuid=module.params['uuid'])
  File "/tmp/ansible_xTG64q/ansible_module_vmware_guest_facts.py", line 128, in getvm
    vm = find_vm_by_id(self.content, vm_id=uuid, vm_id_type="uuid")
  File "/tmp/ansible_xTG64q/ansible_modlib.zip/ansible/module_utils/vmware.py", line 182, in find_vm_by_id
  File "/usr/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py", line 580, in <lambda>
    self.f(*(self.args + (obj,) + args), **kwargs)
  File "/usr/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py", line 385, in _InvokeMethod
    list(map(CheckField, info.params, args))
  File "/usr/lib/python2.7/site-packages/pyVmomi/VmomiSupport.py", line 915, in CheckField
    raise TypeError('Required field "%s" not provided (not @optional)' % info.name)
TypeError: Required field "uuid" not provided (not @optional)
```

##### Potential solution
Poking around in the vmware module and in pyvmomi docs I see the following:
In vmware.py there is this:
```
def find_vm_by_id(content, vm_id, vm_id_type="vm_name", datacenter=None, cluster=None):
    """ UUID is unique to a VM, every other id returns the first match. """
    si = content.searchIndex
    vm = None

    if vm_id_type == 'dns_name':
        vm = si.FindByDnsName(datacenter=datacenter, dnsName=vm_id, vmSearch=True)
    elif vm_id_type == 'inventory_path':
        vm = si.FindByInventoryPath(inventoryPath=vm_id)
        if isinstance(vm, vim.VirtualMachine):
            vm = None
    elif vm_id_type == 'uuid':
        vm = si.FindByUuid(datacenter=datacenter, instanceUuid=vm_id, vmSearch=True)
    elif vm_id_type == 'ip':
        vm = si.FindByIp(datacenter=datacenter, ip=vm_id, vmSearch=True)
    elif vm_id_type == 'vm_name':
        folder = None
        if cluster:
            folder = cluster
        elif datacenter:
            folder = datacenter.hostFolder
        vm = find_vm_by_name(content, vm_id, folder)

    return vm
```

Notice that if vm_id_type == 'uuid' the actual uuid is passed to the instanceUuid parameter and no uuid parameter is set when it calls si.FindByUuid

From the pyvmomi docs for SearchIndex I see the following:
https://github.com/vmware/pyvmomi/blob/master/docs/vim/SearchIndex.rst

```
FindByUuid(datacenter, uuid, vmSearch, instanceUuid):

        Finds a virtual machine or host by BIOS or instance UUID.

    Privilege:
        System.View
    Args:

        datacenter (vim.Datacenter, optional):
            If specified, restricts the query to entities in a particular datacenter. If not specified, the entire inventory is searched.
        uuid (str):
            The UUID to find. If vmSearch is true, the uuid can be either BIOS or instance UUID.
        vmSearch (bool):
            If true, search for virtual machines, otherwise search for hosts.
        instanceUuid (bool, optional, since vSphere API 4.0 ):
            Should only be set when vmSearch is true. If specified, search for virtual machines whose instance UUID matches the given uuid. Otherwise, search for virtual machines whose BIOS UUID matches the given uuid.
```

Notice that instanceUuid is a boolean not the actual UUID. The actual uuid is passed to the uuid parameter and instanceUuid is set to true to indicate searching vs the instance uuid instead of the bios uuid.

If I change vmware.py's line to: 
`vm = si.FindByUuid(datacenter=datacenter, uuid=vm_id, instanceUuid=True, vmSearch=True)`

It works as expected.

I could be using the wrong version of pyvmomi?

**Comment by ansibot** [comment]
cc @nerzhul
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Akasurde** [comment]
I can reproduce this on my following setup -

RHEL 7.3
Python 2.7.5
ESXi 6 
pyvmomi==6.5.0.2017.5.post1
Ansible 2.4.0

**Comment by nerzhul** [comment]
cc @jctanner i think we did similar fix on vmware_guest module recently, no if yes we can backport the fix to this module too

**Comment by Akasurde** [comment]
Similar traceback is seen [here](https://github.com/ansible/ansible/pull/24420#issuecomment-300530167). @nerzhul Are you referring PR #24420 ?

**Comment by Akasurde** [comment]
Interestingly, there is another issue reported - #24835

**Comment by Akasurde** [comment]
resolved_by_pr #26040
